### PR TITLE
Convert data visualizer to Typescript

### PR DESCRIPTION
**Pull request by jeffsieu**
### Summary of changes

- List visualizer code converted to Typescript
- Removed from public/externalLibs
- Added to src/features/ListVisualizer

### Previous implementation
listVisualizer.js exposed some methods to the React side.
1. listVisualizer.js loaded by public/externalLibs/index.js, adds some functions to the `window` instance
2. listVisualizer.js adds a container into the main DOM, to add Konva stages (canvases) into.
2. JsSlangHelper.ts executes listVisualizer.js when `draw_data` is called
3. listVisualizer.js parses the data object and generates a Konva Layer representing the data object.
4. The layer is added in the DOM by listVisualizer.js.
5. listVisualizer.js also maintains the state of the visualizer (current step), and provides functions to increment/decrement step.

### Motivation
This approach overloads the list visualizer with data structure parsing, state maintaining, and DOM management. Ideally, the last two should be handled by React, or at least be separated out in some way. Converting the code to Typescript and separating each class out into their own files will also help make maintaining it easier.

### Current implementation
ListVisualizer.tsx acts as the public-facing class that exposes the `drawData`, `clear` functions.

1. ListVisualizer is a singleton class that maintains a single instance of itself which represents the one generated by pressing the "Run" button. Pressing "Run" again will generate a new instance.
2. JsSlangHelper routes any call to draw_data to ListVisualizer.drawData, which redirects it to the instance function `addStep`.
3. As each draw_data call within a Source file should be a single step, `addStep` encapsulates this function. It creates a new drawing for every structure in the call (as draw_data is a varargs function) and creates a new `Step` out of them. A Step represents a single draw_data call (and is simply an array of drawings in TypeScript).
4. `createDrawing` handles the drawing for each data structure. If it is a pair, it is passed into `Tree`, which has functions responsible for parsing the tree. The tree is parsed so that same objects will not be drawn twice (consider the case where we have an object `a`, and we do `draw_data(list(a, a))`.
5. The parsed tree (a `Tree` instance) is fed into `TreeDrawer`, which then re-traverses the tree and creates Konva objects representing boxes and arrows appropriately. These objects are added into an array as the traversal happens. At the end of the traversal, it wraps all the objects in a `Layer` and returns this.
6. The returned `Layer` object is then forms part of the `Step` that was mentioned earlier.
7. The array of steps is then updated back to the React side content component of list visualizer via the function passed in when the ListVisualizer was `init`ed.
8. The React side content component then renders these steps as children. The current step state is handled by this component as well, as opposed to the ListVisualizer file previously

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update
- [x] Code quality improvements

### Checklist

- [x] I have tested this code
- [x] I have updated the documentation

### CODE REVIEW

**Comment by martin-henz** [comment]
CI seems to complain:
https://github.com/source-academy/cadet-frontend/pull/1685/checks?check_run_id=2275747678

Is that coming from linting?

**Comment by martin-henz** [comment]
Now there are failing test cases:
https://github.com/source-academy/cadet-frontend/pull/1685/checks?check_run_id=2275965995

Make sure you test well. Please mark the PR as "Draft" until it's ready for review.

File: src/features/listVisualizer/tree/Tree.tsx [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+ *  A tree object built based on a list or pair.
+ */
+export class Tree {
+  private _rootNode: PairTreeNode;
+  private nodes: DrawableTreeNode[];
+
+  /**
+   * Constructs a tree given a root node and a list of nodes.
+   * @param {PairTreeNode} rootNode The root node of the tree.
+   * @param {DrawableTreeNode[]} nodes The memoized nodes of the tree in list form.
```
Comment (Anonymxtrix) [inline src/features/listVisualizer/tree/Tree.tsx:20]: The `{PairTreeNode}` and `{DrawableTreeNode[]}` in the function documentation doesn't really do anything

File: src/features/listVisualizer/tree/Tree.tsx [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+  /**
+   * The root node of the tree.
+   */
+  get rootNode(): PairTreeNode {
+    return this._rootNode;
+  }
+
+  /**
+   * Returns the memoized node of the given id.
+   * @param {number} id The id of the node.
```
Comment (Anonymxtrix) [inline src/features/listVisualizer/tree/Tree.tsx:36]: The `{number}` in the function documentation doesn't really do anything

File: src/features/listVisualizer/tree/Tree.tsx [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+  }
+
+  static fromSourceTree(tree: Pair): Tree {
+    let nodeCount = 0;
+
+    /**
+     * Returns a node representing the given tree as a pair.
+     * Also memoizes the pair object, for the case where the
+     * pair appears multiple times in the data structure.
+     * @param {Pair} tree The Source tree to construct a node for.
```
Comment (Anonymxtrix) [inline src/features/listVisualizer/tree/Tree.tsx:49]: The `{Pair}` in the function documentation doesn't really do anything

File: src/features/listVisualizer/tree/Tree.tsx [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+      }
+
+      return node;
+    }
+
+    /**
+     * Returns a node representing the given function.
+     * Also memoizes the function object, for the case where the
+     * function appears multiple times in the data structure.
+     * @param {Function} func The function to construct a node for.
```
Comment (Anonymxtrix) [inline src/features/listVisualizer/tree/Tree.tsx:90]: The `{Function}` in the function documentation doesn't really do anything

File: src/features/listVisualizer/tree/Tree.tsx [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+   *  Draws the box for the pair representing the tree, then recursively draws its children.
+   *  A slash is drawn for empty lists.
+   *
+   *  If a child node has been drawn previously, an arrow is drawn pointing to the children,
+   *  instead of drawing the child node again.
+   * @param {TreeNode} node The node to draw.
+   * @param {number} x The x position to draw at.
+   * @param {number} y The y position to draw at.
+   * @param {number} parentX The x position of the parent. If there is no parent, it is the same as x.
+   * @param {number} parentY The y position of the parent. If there is no parent, it is the same as y.
```
Comment (Anonymxtrix) [inline src/features/listVisualizer/tree/Tree.tsx:162]: The `{TreeNode}` and `{number}` in the function documentation doesn't really do anything

File: src/features/listVisualizer/tree/TreeNode.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import { TreeNode } from './BaseTreeNode';
+import { DataTreeNode } from './DataTreeNode';
+import { DrawableTreeNode } from './DrawableTreeNode';
+import { FunctionTreeNode } from './FunctionTreeNode';
+import { PairTreeNode } from './PairTreeNode';
+
+export { DataTreeNode, DrawableTreeNode, FunctionTreeNode, PairTreeNode, TreeNode };
```
Comment (Anonymxtrix) [inline src/features/listVisualizer/tree/TreeNode.ts:7]: Consider `export { a, b as c } from '...';` for shorter syntax, but its totally up to you haha

**Review by Anonymxtrix — COMMENTED** [review]
Please look through the comments I left. If you don't agree with them or find something better, we can have a discussion. 
👍👍👍 
@jeffsieu @VimuthM

File: src/features/listVisualizer/ListVisualizerUtils.ts [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+export function isPair(data: Data): data is Pair {
+  return is_pair(data);
+}
+
+export function isList(data: Data): data is List {
+  return is_list(data);
+}
+
+export function isNull(data: Data): data is EmptyList {
```
[the remark this answers is not in the document]
Comment (jeffsieu) [reply src/features/listVisualizer/ListVisualizerUtils.ts:103]: Ah, it checks if it is an empty Source list, so in this case I suppose it should be `isEmptyList`

**Comment by martin-henz** [comment]
CI complaining about linting, I think. Please run 
yarn format
and push the diff.

File: src/features/listVisualizer/tree/BaseTreeNode.ts [hunk 8]
```diff
@@ -0,0 +1,9 @@
+export class TreeNode {
+  public left: TreeNode | number | null;
+  public right: TreeNode | number | null;
+
+  constructor() {
+    this.left = null;
+    this.right = null;
+  }
+}
```
[the remark this answers is not in the document]
Comment (jeffsieu) [reply src/features/listVisualizer/tree/BaseTreeNode.ts:9]: Hi, what do you mean by this? How may we use generics in this case?

File: src/commons/sideContent/SideContentListVisualizer.tsx [hunk 9]
```diff
@@ -16,15 +19,26 @@ const listVisualizerKeyMap = {
...
   constructor(props: any) {
     super(props);
-    this.state = { loading: true };
-  }
+    // ListVisualizer.clear();
+    this.state = { steps: [], currentStep: 0 };
+    ListVisualizer.init(steps => {
+      if (!steps) {
+        //  Blink icon
+        const icon = document.getElementById('data_visualiser-icon');
```
[the remark this answers is not in the document]
Comment (jeffsieu) [reply src/commons/sideContent/SideContentListVisualizer.tsx:33]: Hmm? This is the icon for the data visualizer in the side content panel. I think the name `data_visualizer-icon` would be more appropriate.

File: src/features/listVisualizer/tree/BaseTreeNode.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (Anonymxtrix) [reply src/features/listVisualizer/tree/BaseTreeNode.ts:9]: No its just more fyi hahaha

**Review by Anonymxtrix — APPROVED** [review]

**Review by MarcusTXK — APPROVED** [review]
With the most recent updates, the tests and formatting have been fixed. Tested on SourceAcademy locally and seems to work well without any bugs.

**Pull request merged**